File: 17.cs
using System;

class Entity {
  static int nextSerialNo;
  int serialNo;
  public Entity() {
    serialNo = nextSerialNo++;
  }
  public int GetSerialNo() { 
    return serialNo;
  }
  public static int GetNextSerialNo() { 
    return nextSerialNo;
  }
  public static void SetNextSerialNo(int value) { 
    nextSerialNo = value;
  } 
}

class Test {
  static void Main() { 
    Entity.SetNextSerialNo(1000);
    Entity e1 = new Entity(); 
    Entity e2 = new Entity();
    Console.WriteLine(e1.GetSerialNo()); // 1000
    Console.WriteLine(e2.GetSerialNo()); // 1001
    Console.WriteLine(Entity.GetNextSerialNo());  //1002
  }
}

File: 17.cpp
#include <iostream>

// issue : static field initialization
class Entity {
  public:
    static int nextSerialNo;
    int serialNo;
    Entity() {
      serialNo = nextSerialNo++;
    }
    int GetSerialNo() { 
      return serialNo;
    }
    static int GetNextSerialNo() { 
      return nextSerialNo;
    }
    static void SetNextSerialNo(int value) { 
      nextSerialNo = value;
    } 
};

using namespace std;

int Entity::nextSerialNo = 0;

int main() { 
  Entity::SetNextSerialNo(1000);
  Entity *e1 = new Entity(); 
  Entity *e2 = new Entity();
  cout << e1->GetSerialNo() << endl; // 1000
  cout << e2->GetSerialNo() << endl; // 1001
  cout << Entity::GetNextSerialNo() << endl;  //1002
}
